lib/ecto/query/planner.ex

```diff
@@ -1682,6 +1718,31 @@ defmodule Ecto.Query.Planner do
     end
   end
 
+  @spec assert_same_types!(Ecto.Query.t, left :: Macro.t, right :: Ecto.SubQuery.t) :: true
+  defp assert_same_types!(query, left, right) do
```

Comment:
Since this is subquery specific, let's call it `assert_in_subquery_type!` or similar.